File: 1163. Last Substring in Lexicographical Order.py
class Solution:
    def lastSubstring(self, s: str) -> str:
        mx = max(s)
        p = [i for i,c in enumerate(s) if c==mx]
        ans = p[-1]
        for i in reversed(range(len(p)-1)):
            x, y = p[i], ans
            while x<ans and y<len(s) and s[x]==s[y]:
                x, y = x+1, y+1
            if x==ans or y==len(s) or s[x]>s[y]:
                ans = p[i]
        return s[ans:]

File: 1163. Last Substring in Lexicographical Order.cpp
class Solution {
public:
    string lastSubstring(string s) {
        char mx = 0;
        vector<int> p;
        for (int i=0; i<s.size(); ++i)
            if (s[i] > mx) {
                mx = s[i];
                p = {i};
            } else if (s[i] == mx)
                p.push_back(i);
        int ans = p.back();
        for (int i=(int)p.size()-2; i>=0; --i) {
            int x=p[i], y=ans;
            for (;x<ans && y<s.size() && s[x]==s[y]; ++x,++y);
            if (x==ans || y==s.size() || s[y] < s[x]) ans = p[i];
        }
        return s.substr(ans);
    }
};